### PR TITLE
THOR Dashboard State Flow

### PR DESCRIPTION
**Pull request by arciisine**
Rework the state flow to allow for single path for all state changes, which will allow for url synchronization

### CODE REVIEW

File: src/app/dashboard/dashboard.component.ts [hunk 1]
```diff
@@ -183,84 +166,97 @@ export class DashboardComponent implements AfterViewInit, OnInit, OnDestroy {
...
+            favicon.setAttribute('rel', rel);
+            favicon.setAttribute('type', 'image/x-icon');
+            favicon.setAttribute('href', icon);
+            head.appendChild(favicon);
+        }
+    }
 
-        title.innerText = this.projectTitle;
+    @DashboardModified()
+    private generalChange() {
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/dashboard/dashboard.component.ts:250]: This function is just here to trigger DashboardModified, right?

File: src/app/models/neon-namespaces.ts [hunk 2]
```diff
@@ -15,10 +15,7 @@
 
 export namespace neonEvents {
     export const DASHBOARD_ERROR = 'DASHBOARD_ERROR';
-    export const DASHBOARD_READY = 'DASHBOARD_READY';
     export const DASHBOARD_REFRESH = 'DASHBOARD_REFRESH';
-    export const DASHBOARD_RESET = 'DASHBOARD_RESET';
-    export const DASHBOARD_STATE = 'DASHBOARD_STATE';
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/models/neon-namespaces.ts:21]: Please remove these events from the dashboard event guide:  https://github.com/NextCenturyCorporation/neon-dash-internal/blob/lorelei-master/DASHBOARD_EVENT_GUIDE.md

File: src/app/services/config.service.ts [hunk 3]
```diff
@@ -18,21 +18,29 @@ import * as yaml from 'js-yaml';
...
 export class ConfigService {
     private configErrors = [];
-    private source = new ReplaySubject<NeonConfig>(1);
+    private source = new Subject<NeonConfig>();
 
     $source: Observable<NeonConfig>;
 
-    static as(config: NeonConfig) {
-        return new ConfigService(null, null).setActive(config);
+    static as(config: NeonConfig | null) {
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/services/config.service.ts]: The "as" function is just used for testing, so it seems weird to me to include it in the service itself.  I'd hate for someone to incorrectly use it in non-test code.

File: src/app/services/dashboard.service.ts [hunk 4]
```diff
@@ -13,63 +13,74 @@
...
+        private searchService: AbstractSearchService
+    ) {
+        this.configSource = this.configService
+            .getActive()
+            .pipe(
+                mergeMap((config) => this.onConfigChange(config)),
+                shareReplay(1)
+            );
+        this.stateSource = this.stateSubject
+            .pipe(shareReplay(1));
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/services/dashboard.service.ts:60]: Forgive my ignorance, but what's the difference between stateSubject and stateSource here?  Why is it useful to have both?

**Comment by ThomasSchellenbergNextCentury** [comment]
@arciisine I'm getting lint errors running `npm run-script lint`

**Comment by ThomasSchellenbergNextCentury** [comment]
@arciisine For our meeting this Friday on Observables, it would be great if you could speak about the RxJS functions you're using here, particularly in the Config and Dashboard Services, and how they work.  While we could all read the documentation ourselves, I think that having a short overview together as a team would also be nice.

File: src/app/dashboard/dashboard.component.ts (hunk 1 above)
Comment (arciisine) [reply src/app/dashboard/dashboard.component.ts:250]: Yes, its a catchall for any method that needs to modify the dashboard but doesn't have explicit functionailty

File: src/app/services/dashboard.service.ts (hunk 4 above)
Comment (arciisine) [reply src/app/services/dashboard.service.ts:60]: Subject is writable, and is private as only the service should be able to emit events.  The source/observable is public and is read only.

File: src/app/services/config.service.ts (hunk 3 above)
Comment (arciisine) [reply src/app/services/config.service.ts]: Sounds like a good idea.

Comment (arciisine) [reply src/app/services/config.service.ts]: Resolved, had to touch quite a bit, hopefully made it less brittle for the future.

**Comment by arciisine** [comment]
Lint errors should be resolved, and all feedback should be addressed.

**Review by ThomasSchellenbergNextCentury — APPROVED** [review]

**Pull request merged**